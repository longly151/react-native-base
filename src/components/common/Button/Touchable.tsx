import React from 'react';
import {
  Platform,
  StyleSheet,
  TouchableNativeFeedback,
  TouchableNativeFeedbackProps,
  TouchableOpacity,
  TouchableOpacityProps,
  StyleProp,
  ViewStyle,
} from 'react-native';
import withTailwind from '@components/hoc/withTailwind';
import View from '../View/View';
import { omit, pick } from 'lodash';

/* istanbul ignore next */
const Component: any = Platform.OS === 'android' ? TouchableNativeFeedback : TouchableOpacity;

export interface TouchableProps extends TouchableNativeFeedbackProps, TouchableOpacityProps {
  className?: string;
}

export const containerStyleKeys: any = {
  android: [
    'flex',
    'flexGrow',
    'flexShrink',
    'alignSelf',
    'margin',
    'marginLeft',
    'marginRight',
    'marginTop',
    'marginBottom',
    'marginVertical',
    'marginHorizontal',
    'borderRadius',
    'borderTopLeftRadius',
    'borderTopRightRadius',
    'borderBottomLeftRadius',
    'borderBottomRightRadius',
    'borderWidth',
    'borderTopWidth',
    'borderBottomWidth',
    'borderLeftWidth',
    'borderRightWidth',
    'borderColor',
    'borderTopColor',
    'borderBottomColor',
    'borderLeftColor',
    'borderRightColor',
    'backgroundColor',
    'elevation',
    'position',
  ],
  ios: [
    'flex',
    'flexGrow',
    'flexShrink',
    'alignSelf',
    'margin',
    'marginLeft',
    'marginRight',
    'marginTop',
    'marginBottom',
    'marginVertical',
    'marginHorizontal',
    'elevation',
    'position',
  ],
};

const Touchable: React.FC<TouchableProps> = React.forwardRef((props, ref: any) => {
  const { style: styleProp, ...otherProps } = props;

  const style: StyleProp<ViewStyle> = StyleSheet.flatten(styleProp);

  const otherStyle: StyleProp<ViewStyle> = omit(style, containerStyleKeys[Platform.OS]);

  const borderWidth =
    style?.borderWidth ||
    style?.borderLeftWidth ||
    style?.borderRightWidth ||
    style?.borderTopWidth ||
    style?.borderBottomWidth;

  const containerStyle: StyleProp<ViewStyle> = {
    ...pick(style, containerStyleKeys[Platform.OS]),
    justifyContent: 'center',
    overflow: Platform.OS === 'android' ? 'hidden' : 'visible',
    width:
      typeof otherStyle.width === 'number' && borderWidth
        ? otherStyle.width + borderWidth
        : otherStyle.width,
    height:
      typeof otherStyle.height === 'number' && borderWidth
        ? otherStyle.height + borderWidth
        : otherStyle.height,
  };

  return (
    <View testID="ContainerView" style={containerStyle}>
      <Component
        testID="Touchable"
        ref={ref}
        {...otherProps}
        // style={{ flexGrow: otherStyle.flexGrow }}
      >
        <View testID="TouchableView" style={otherStyle}>
          {props.children}
        </View>
      </Component>
    </View>
  );
});

export interface TouchableType
  extends Omit<
      TouchableNativeFeedback,
      | 'UNSAFE_componentWillReceiveProps'
      | 'UNSAFE_componentWillUpdate'
      | 'componentDidUpdate'
      | 'componentWillReceiveProps'
      | 'componentWillUpdate'
      | 'getSnapshotBeforeUpdate'
      | 'props'
      | 'refs'
      | 'setState'
      | 'shouldComponentUpdate'
    >,
    TouchableOpacity {}

export default withTailwind(Touchable) as React.FC<
  TouchableProps & { ref?: React.LegacyRef<TouchableNativeFeedback | TouchableOpacity> }
>;
