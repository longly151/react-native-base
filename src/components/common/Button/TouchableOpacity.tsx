import React from 'react';
import {
  StyleProp,
  TouchableOpacity as RNTouchableOpacity,
  TouchableOpacityProps as RNTouchableOpacityProps,
  ViewStyle,
} from 'react-native';
import withTailwind from '@components/hoc/withTailwind';

export interface TouchableOpacityProps extends RNTouchableOpacityProps {
  className?: string;
}

const TouchableOpacity: React.FC<TouchableOpacityProps> = React.forwardRef((props, ref: any) => {
  const { style: styleProp, ...otherProps } = props;
  const style: StyleProp<ViewStyle> = [styleProp];
  return (
    <RNTouchableOpacity testID="TouchableOpacity" ref={ref} {...otherProps} style={style}>
      {props.children}
    </RNTouchableOpacity>
  );
});

export interface TouchableOpacityType extends RNTouchableOpacity {}

export default withTailwind(TouchableOpacity) as React.FC<
  TouchableOpacityProps & { ref?: React.LegacyRef<RNTouchableOpacity> }
>;
