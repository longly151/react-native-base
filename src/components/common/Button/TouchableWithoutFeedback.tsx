import React from 'react';
import {
  StyleProp,
  TouchableWithoutFeedback as RNTouchableWithoutFeedback,
  TouchableWithoutFeedbackProps as RNTouchableWithoutFeedbackProps,
  ViewStyle,
} from 'react-native';
import withTailwind from '@components/hoc/withTailwind';

export interface TouchableWithoutFeedbackProps extends RNTouchableWithoutFeedbackProps {
  className?: string;
  children?: any;
}

const TouchableWithoutFeedback: React.FC<TouchableWithoutFeedbackProps> = React.forwardRef(
  (props, ref: any) => {
    const { style: styleProp, ...otherProps } = props;
    const style: StyleProp<ViewStyle> = [styleProp];
    return (
      <RNTouchableWithoutFeedback
        testID="TouchableWithoutFeedback"
        ref={ref as any}
        {...otherProps}
        style={style}
      >
        {props.children}
      </RNTouchableWithoutFeedback>
    );
  },
);

export interface TouchableWithoutFeedbackType extends RNTouchableWithoutFeedback {}

export default withTailwind(TouchableWithoutFeedback) as React.FC<
  TouchableWithoutFeedbackProps & { ref?: React.LegacyRef<RNTouchableWithoutFeedback> }
>;
