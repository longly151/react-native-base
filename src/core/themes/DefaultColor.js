module.exports = {
  dark: false,
  colors: {
    primary: '#219EE5',
    background: '#f2f2f7',
    card: '#ffffff',
    text: '#000000',
    border: '#d8d8d8',
    notification: '#ff3b30',
    grey0: '#393e42',
    grey1: '#43484d',
    grey2: '#5e6977',
    grey3: '#86939e',
    grey4: '#bdc6cf',
    grey5: '#e1e8ee',
    white: '#ffffff',
    black: '#242424',
    info: '#36A3F7',
    success: '#34BFA3',
    warning: '#FFB822',
    error: '#F4516C',
  },
};
