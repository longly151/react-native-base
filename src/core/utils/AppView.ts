import { Font } from '@fonts';
import DefaultColor from '@themes/DefaultColor';
import { Dimensions, StyleProp, ViewStyle, TextStyle, ColorValue, ScaledSize } from 'react-native';
import ViewConfig from '@themes/ViewConfig';

const { width, height } = Dimensions.get('window');

/**
 * Shadow
 */
const androidDepth = {
  umbra: [
    '0px 2px 1px -1px',
    '0px 3px 1px -2px',
    '0px 3px 3px -2px',
    '0px 2px 4px -1px',
    '0px 3px 5px -1px',
    '0px 3px 5px -1px',
    '0px 4px 5px -2px',
    '0px 5px 5px -3px',
    '0px 5px 6px -3px',
    '0px 6px 6px -3px',
    '0px 6px 7px -4px',
    '0px 7px 8px -4px',
    '0px 7px 8px -4px',
    '0px 7px 9px -4px',
    '0px 8px 9px -5px',
    '0px 8px 10px -5px',
    '0px 8px 11px -5px',
    '0px 9px 11px -5px',
    '0px 9px 12px -6px',
    '0px 10px 13px -6px',
    '0px 10px 13px -6px',
    '0px 10px 14px -6px',
    '0px 11px 14px -7px',
    '0px 11px 15px -7px',
  ],
  penumbra: [
    '0px 1px 1px 0px',
    '0px 2px 2px 0px',
    '0px 3px 4px 0px',
    '0px 4px 5px 0px',
    '0px 5px 8px 0px',
    '0px 6px 10px 0px',
    '0px 7px 10px 1px',
    '0px 8px 10px 1px',
    '0px 9px 12px 1px',
    '0px 10px 14px 1px',
    '0px 11px 15px 1px',
    '0px 12px 17px 2px',
    '0px 13px 19px 2px',
    '0px 14px 21px 2px',
    '0px 15px 22px 2px',
    '0px 16px 24px 2px',
    '0px 17px 26px 2px',
    '0px 18px 28px 2px',
    '0px 19px 29px 2px',
    '0px 20px 31px 3px',
    '0px 21px 33px 3px',
    '0px 22px 35px 3px',
    '0px 23px 36px 3px',
    '0px 24px 38px 3px',
  ],
  ambient: [
    '0px 0px 0px 0px',
    '0px 1px 3px 0px',
    '0px 1px 5px 0px',
    '0px 1px 8px 0px',
    '0px 1px 10px 0px',
    '0px 1px 14px 0px',
    '0px 1px 18px 0px',
    '0px 2px 16px 1px',
    '0px 3px 14px 2px',
    '0px 3px 16px 2px',
    '0px 4px 18px 3px',
    '0px 4px 20px 3px',
    '0px 5px 22px 4px',
    '0px 5px 24px 4px',
    '0px 5px 26px 4px',
    '0px 6px 28px 5px',
    '0px 6px 30px 5px',
    '0px 6px 32px 5px',
    '0px 7px 34px 6px',
    '0px 7px 36px 6px',
    '0px 8px 38px 7px',
    '0px 8px 40px 7px',
    '0px 8px 42px 7px',
    '0px 9px 44px 8px',
    '0px 9px 46px 8px',
  ],
};
function interpolate(i: number, a: number, b: number, a2: number, b2: number) {
  return ((i - a) * (b2 - a2)) / (b - a) + a2;
}
function parseShadow(raw: any) {
  const values = raw.split(' ').map((val: any) => +val.replace('px', ''));
  return {
    x: values[0],
    y: values[1],
    blur: values[2],
    spread: values[3], // unused
  };
}
function onDepthChange(depth: number) {
  const s = parseShadow(androidDepth.penumbra[depth]);
  const y = s.y === 1 ? 1 : Math.floor(s.y * 0.5);
  return {
    elevation: depth + 1, // Android
    // iOS
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: y,
    },
    shadowOpacity: Number.parseFloat(interpolate(depth, 1, 24, 0.2, 0.6).toFixed(2)),
    shadowRadius: Number.parseFloat(interpolate(s.blur, 1, 38, 1, 16).toFixed(2)),
  };
}

class CAppView {
  private static _instance: CAppView;

  private constructor() {
    // ...
  }

  public static get Instance(): CAppView {
    /* istanbul ignore else */
    if (!this._instance) {
      this._instance = new this();
    }
    return CAppView._instance;
  }

  fontFamily: Font = 'GoogleSans-Regular';

  safeAreaInsets = {
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  };

  shadow = (depth: number = 10) => (depth >= 1 ? onDepthChange(depth - 1) : {});

  bottomNavigationBarHeight = 45;

  headerHeight = 50;

  bodyHeight =
    height -
    this.headerHeight -
    this.bottomNavigationBarHeight -
    this.safeAreaInsets.top -
    this.safeAreaInsets.bottom;

  bodyPaddingHorizontal = Number.parseFloat(ViewConfig.spacing.DEFAULT);

  bodyWidth = width - 2 * this.bodyPaddingHorizontal;

  roundedBorderRadius = Number.parseFloat(ViewConfig.borderRadius.DEFAULT);

  bottomSheetBorderRadius = Number.parseFloat(ViewConfig.borderRadius.card);

  itemMarginHorizontal = 10;

  screenWidth = width;

  screenHeight = height;

  isHorizontal = this.screenWidth > this.screenHeight;

  // Carousel
  sliderWidth = this.screenWidth;

  itemWidth = (this.isHorizontal ? 0.35 : 0.75) * this.screenWidth + this.itemMarginHorizontal * 2;

  itemHeight = 300;

  carouselBorderRadius = 8;

  onDimensionChange = (dimensions: { window: ScaledSize; screen: ScaledSize }) => {
    const { window } = dimensions;
    this.screenWidth = window.width;
    this.screenHeight = window.height;
    const isHorizontal = window.width > window.height;
    this.isHorizontal = isHorizontal;

    if (isHorizontal) {
      // Carousel
      this.itemWidth = 0.35 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    } else {
      this.itemWidth = 0.75 * window.width + this.itemMarginHorizontal * 2;
      this.sliderWidth = window.width;
    }
  };

  initSafeArea = (safeAreaInsets: { top: number; bottom: number; left: number; right: number }) => {
    const dimensions = Dimensions.get('window');
    this.safeAreaInsets = safeAreaInsets;
    this.bodyHeight =
      dimensions.height -
      this.headerHeight -
      this.bottomNavigationBarHeight -
      safeAreaInsets.top -
      safeAreaInsets.bottom;
  };

  getHeaderStyle = (theme: typeof DefaultColor) =>
    ({
      headerStyle: {
        backgroundColor: theme.colors.background,
      } as StyleProp<ViewStyle>,
      headerTitleStyle: {
        fontSize: 22,
        color: theme.colors.grey0,
        fontWeight: 'bold',
        fontFamily: AppView.fontFamily,
      } as StyleProp<TextStyle>,
      headerShadowVisible: false,
      headerBackTitleVisible: false,
      headerBackButtonMenuEnabled: false,
      headerTintColor: theme.colors.text,
    } as any);

  getQuickStyle = ({
    outline,
    clear,
    backgroundColor,
    textColor,
    borderColor,
  }: {
    outline?: boolean;
    clear?: boolean;
    backgroundColor: ColorValue;
    textColor: ColorValue;
    borderColor?: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    let quickTextStyles: StyleProp<TextStyle> = {};
    if (outline) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderColor: borderColor || backgroundColor,
        borderWidth: 1,
        elevation: 0,
      };
      quickTextStyles = {
        color: textColor || backgroundColor,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        elevation: 0,
        borderRadius: 0,
      };
      quickTextStyles = { color: backgroundColor };
    } else {
      quickStyles = {
        backgroundColor,
      };
    }
    return { quickStyles, quickTextStyles };
  };

  getInputQuickStyle = ({
    outline,
    clear,
  }: {
    outline?: boolean;
    clear?: boolean;
    textColor: ColorValue;
  }) => {
    let quickStyles: StyleProp<ViewStyle> = {};
    if (outline) {
      quickStyles = {
        borderWidth: 1,
        elevation: 0,
      };
    } else if (clear) {
      quickStyles = {
        backgroundColor: 'transparent',
        borderRadius: 0,
        borderBottomWidth: 1,
        elevation: 0,
      };
    } else {
      quickStyles = {
        borderWidth: 1,
      };
    }
    return { quickStyles };
  };
}

const AppView = CAppView.Instance;
export default AppView;
