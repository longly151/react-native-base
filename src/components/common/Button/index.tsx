import React, { useMemo } from 'react';
import {
  StyleProp,
  Button as RNButton,
  ButtonProps as RNButtonProps,
  ViewStyle,
  TextStyle,
  ColorValue,
  ActivityIndicator,
  Text,
  TextProps,
  StyleSheet,
} from 'react-native';
import withTailwind from '@components/hoc/withTailwind';
import { useTheme } from '@react-navigation/native';
import AppView from '@utils/AppView';
import { ViewProps } from '../View/View';
import Icon, { IconProps } from '../Icon';
import Touchable, { TouchableProps } from './Touchable';
import Color from 'color';
import CONSTANT from '@configs/constant';
import TouchableOpacity from './TouchableOpacity';

export interface ButtonProps
  extends Omit<RNButtonProps, 'color' | 'title'>,
    ViewProps,
    TouchableProps {
  title?: string;
  leftIconProps?: IconProps;
  rightIconProps?: IconProps;
  textStyle?: StyleProp<TextStyle>;
  textProps?: TextProps;
  backgroundColor?: ColorValue;
  loading?: boolean;
  outline?: boolean;
  clear?: boolean;
  style?: StyleProp<ViewStyle & { color?: ColorValue | undefined }>;
  touchComponent?: 'TouchableOpacity' | 'Touchable';
}

const Button: React.FC<ButtonProps> = React.forwardRef((props, ref: any) => {
  const theme = useTheme();
  const {
    style: rawStyleProp,
    disabled: disabledProp,
    accessibilityState: accessibilityStateProp,
    title,
    leftIconProps,
    rightIconProps,
    textStyle: rawTextStyleProp,
    textProps,
    backgroundColor: backgroundColorProp,
    loading,
    outline,
    clear,
    touchComponent,
    ...otherProps
  } = props;
  const styleProp = StyleSheet.flatten(rawStyleProp);
  const textStyleProp = StyleSheet.flatten(rawTextStyleProp);

  // @ts-ignore
  const fontFamily = styleProp?.fontFamily || AppView.fontFamily;
  const backgroundColor = backgroundColorProp || theme.colors.primary;
  const borderColor =
    styleProp?.borderColor ||
    styleProp?.borderTopColor ||
    styleProp?.borderBottomColor ||
    styleProp?.borderLeftColor ||
    styleProp?.borderRightColor;
  const textColor =
    styleProp?.color ||
    textStyleProp?.color ||
    (outline || clear ? backgroundColor || borderColor || 'white' : 'white');

  const { quickStyles, quickTextStyles } = AppView.getQuickStyle({
    outline,
    clear,
    backgroundColor,
    textColor,
    borderColor,
  });

  const buttonStyles: StyleProp<ViewStyle> = [
    {
      elevation: 4,
      borderRadius: AppView.roundedBorderRadius,
      flexDirection: 'row',
      justifyContent: 'center',
    },
    quickStyles,
    styleProp,
  ];

  const defaultTextStyles: StyleProp<TextStyle> = {
    textAlign: 'center',
    fontFamily,
    margin: CONSTANT.BUTTON.MARGIN,
    fontSize: CONSTANT.BUTTON.FONT_SIZE,
    color: textColor,
    fontWeight: CONSTANT.BUTTON.FONT_WEIGHT as TextStyle['fontWeight'],
    alignSelf: 'center',
    opacity: loading ? 0 : 1,
    textTransform: (styleProp as any)?.textTransform,
  };
  const textStyles: StyleProp<TextStyle> = [defaultTextStyles, quickTextStyles, textStyleProp];

  const {
    color: iconColor,
    fontSize: iconSize,
    ...otherTextStyle
  } = StyleSheet.flatten(textStyles);

  // const iconPaddingHorizontal =
  //   otherTextStyle.paddingLeft || otherTextStyle.paddingRight || otherTextStyle.paddingHorizontal;

  const leftIconStyles: StyleProp<TextStyle> = [
    otherTextStyle,
    {
      margin: 0,
      marginRight: title
        ? CONSTANT.BUTTON.ICON_TITLE_MARGIN_HORIZONTAL -
          ((otherTextStyle.marginHorizontal || otherTextStyle.margin || 0) as number)
        : 0,
      marginLeft: title ? otherTextStyle.marginHorizontal || otherTextStyle.margin : 0,
    },
    leftIconProps?.style,
  ];

  const rightIconStyles: StyleProp<TextStyle> = [
    otherTextStyle,
    {
      margin: 0,
      marginLeft: title ? -((otherTextStyle.margin as number) / 2) : 0,
      marginRight: title ? otherTextStyle.margin : 0,
    },
    rightIconProps?.style,
  ];

  const disabled = disabledProp != null ? disabledProp : accessibilityStateProp?.disabled;
  const accessibilityState =
    disabled !== accessibilityStateProp?.disabled
      ? { ...accessibilityStateProp, disabled }
      : accessibilityStateProp;

  /* istanbul ignore else */
  if (disabled) {
    buttonStyles.push({
      elevation: 0,
      backgroundColor: theme.colors.grey5,
    });
    textStyles.push({ color: Color(theme.colors.grey3).lighten(0.2).hex() });
  }

  const TouchComponent = useMemo(
    () => (touchComponent === 'TouchableOpacity' ? TouchableOpacity : Touchable),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  return (
    <TouchComponent
      ref={ref}
      {...otherProps}
      accessibilityState={accessibilityState}
      disabled={disabled || loading}
      style={buttonStyles}
    >
      {leftIconProps ? (
        <Icon
          testID="LeftIcon"
          color={iconColor}
          size={iconSize}
          {...leftIconProps}
          style={leftIconStyles}
        />
      ) : null}
      {title ? (
        <Text testID="ButtonTitleText" {...textProps} style={textStyles}>
          {title}
        </Text>
      ) : null}
      {rightIconProps ? (
        <Icon
          testID="RightIcon"
          color={iconColor}
          size={iconSize}
          {...rightIconProps}
          style={rightIconStyles}
        />
      ) : null}
      {loading ? (
        <ActivityIndicator
          testID="Loading"
          style={{
            position: 'absolute',
            alignSelf: 'center',
            justifyContent: 'center',
            height: '100%',
          }}
          color={disabled ? theme.colors.grey3 : textColor}
        />
      ) : null}
    </TouchComponent>
  );
});

Button.defaultProps = {};

export interface ButtonType extends RNButton {}

export default withTailwind(Button) as React.FC<ButtonProps & { ref?: React.LegacyRef<RNButton> }>;
