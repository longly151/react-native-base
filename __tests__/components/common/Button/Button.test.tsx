import React from 'react';
import { render } from '@utils/TestHelper';
import Button from '@components/common/Button/index';
import DefaultColor from '@themes/DefaultColor';
import { Platform, View } from 'react-native';
import _ from 'lodash';
import Touchable, { containerStyleKeys } from '@components/common/Button/Touchable';
import Color from 'color';
import TouchableOpacity from '@components/common/Button/TouchableOpacity';
import TouchableWithoutFeedback from '@components/common/Button/TouchableWithoutFeedback';
import AppView from '@utils/AppView';

let mockTheme = DefaultColor;
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useTheme: jest.fn().mockImplementation(() => mockTheme),
}));

beforeEach(() => {
  mockTheme = DefaultColor;
  Platform.OS = 'ios';
});

test('should render correctly', () => {
  const { toJSON, getByTestId } = render(<Button />);
  expect(toJSON()).toMatchSnapshot();
  expect(getByTestId('TouchableView')).toBeTruthy();
});

test('should have working props', () => {
  const RootComponent = (
    <Button className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500" />
  );
  const { getByTestId, rerender } = render(RootComponent);

  const style = {
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
  };

  let containerStyle = _.pick(style, containerStyleKeys[Platform.OS]);
  let otherStyle = _.omit(style, containerStyleKeys[Platform.OS]);
  expect(getByTestId('ContainerView')).toHaveStyle(containerStyle);
  expect(getByTestId('TouchableView')).toHaveStyle(otherStyle);

  /**
   * Shadow Button
   */
  const { getByTestId: getByTestIdShadow } = render(<Button style={AppView.shadow(5)} />);

  expect(getByTestIdShadow('ContainerView')).toHaveStyle({ elevation: 5 });
  expect(getByTestIdShadow('TouchableView')).toHaveStyle({
    shadowColor: '#000',
    // shadowOffset: { height: 2, width: 0 },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
  });

  /**
   * Outline Button
   */
  const { getByTestId: getByTestIdOutline } = render(<Button outline />);

  expect(getByTestIdOutline('ContainerView')).toHaveStyle({ elevation: 0 });
  expect(getByTestIdOutline('TouchableView')).toHaveStyle({
    backgroundColor: 'transparent',
    borderWidth: 1,
  });

  /**
   * Clear Button
   */
  const { getByTestId: getByTestIdClear } = render(<Button clear />);

  expect(getByTestIdClear('ContainerView')).toHaveStyle({ elevation: 0 });
  expect(getByTestIdClear('TouchableView')).toHaveStyle({
    backgroundColor: 'transparent',
    borderWidth: undefined,
  });

  /**
   * Icon Button
   */

  rerender(
    <Button
      leftIconProps={{
        type: 'material-community',
        name: 'arrow-left',
      }}
      rightIconProps={{
        type: 'material-community',
        name: 'arrow-right',
      }}
    />,
  );
  expect(getByTestId('LeftIcon')).toBeTruthy();
  expect(getByTestId('RightIcon')).toBeTruthy();

  /**
   * Loading Button
   */
  rerender(<Button loading title="Loading Button" />);
  expect(getByTestId('Loading')).toBeTruthy();
  expect(getByTestId('ButtonTitleText')).toHaveStyle({ opacity: 0 });

  /**
   * Disabled Button
   */
  rerender(<Button disabled title="Disabled Button" />);
  expect(getByTestId('ButtonTitleText')).toHaveStyle({
    color: Color(mockTheme.colors.grey3).lighten(0.2).hex(),
  });
  expect(getByTestId('TouchableView')).toHaveStyle({
    backgroundColor: mockTheme.colors.grey5,
  });

  /**
   * Android
   */
  Platform.OS = 'android';
  const { getByTestId: getByTestIdAndroid } = render(RootComponent);
  containerStyle = _.pick(style, containerStyleKeys[Platform.OS]);
  otherStyle = _.omit(style, containerStyleKeys[Platform.OS]);
  expect(getByTestIdAndroid('ContainerView')).toHaveStyle(containerStyle);
  expect(getByTestIdAndroid('TouchableView')).toHaveStyle(otherStyle);
});

test('should have working props [Touchable]', () => {
  const RootComponent = (
    <Touchable className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500" />
  );
  const { getByTestId } = render(RootComponent);

  const style = {
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
  };

  const containerStyle = _.pick(style, containerStyleKeys[Platform.OS]);
  const otherStyle = _.omit(style, containerStyleKeys[Platform.OS]);
  expect(getByTestId('ContainerView')).toHaveStyle(containerStyle);
  expect(getByTestId('TouchableView')).toHaveStyle(otherStyle);

  /**
   * Android
   */
  Platform.OS = 'android';
  Platform.Version = 31;
  const { getByTestId: getByTestIdAndroid } = render(<Touchable className="w-1 h-1 border" />);

  expect(getByTestIdAndroid('ContainerView')).toHaveStyle({ width: 5, height: 5 });
});

test('should have working props [TouchableOpacity]', () => {
  const RootComponent = (
    <TouchableOpacity className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500" />
  );
  const { getByTestId } = render(RootComponent);

  expect(getByTestId('TouchableOpacity')).toHaveStyle({
    flexGrow: 1,
    flexShrink: 1,
    flexDirection: 'row',
    alignSelf: 'center',
    justifyContent: 'space-between',
    alignItems: 'center',
    borderBottomLeftRadius: 4,
    borderBottomRightRadius: 4,
    borderTopLeftRadius: 4,
    borderTopRightRadius: 4,
    borderLeftWidth: 1,
    borderRightWidth: 1,
    borderTopWidth: 1,
    borderBottomWidth: 1,
    borderLeftColor: 'rgba(3, 105, 161, 1)',
    borderRightColor: 'rgba(3, 105, 161, 1)',
    borderTopColor: 'rgba(3, 105, 161, 1)',
    borderBottomColor: 'rgba(3, 105, 161, 1)',
    backgroundColor: 'rgba(14, 165, 233, 1)',
    marginLeft: 4,
    marginRight: 4,
    marginTop: 4,
    marginBottom: 4,
    paddingLeft: 4,
    paddingRight: 4,
    paddingTop: 4,
    paddingBottom: 4,
  });
});

test('should have working props [TouchableWithoutFeedback]', () => {
  const RootComponent = (
    <TouchableWithoutFeedback className="flex-1 flex-row self-center justify-between items-center rounded border mx-1 my-1 p-1 border-sky-700 bg-sky-500">
      <View />
    </TouchableWithoutFeedback>
  );
  const { queryByTestId } = render(RootComponent);

  expect(queryByTestId('TouchableWithoutFeedback')?.props.style).toBeFalsy();
});
