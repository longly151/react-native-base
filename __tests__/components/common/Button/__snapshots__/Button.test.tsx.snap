// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<View
  style={
    Array [
      Array [
        Object {},
        Object {
          "elevation": 4,
          "height": undefined,
          "justifyContent": "center",
          "overflow": "visible",
          "width": undefined,
        },
      ],
    ]
  }
  testID="ContainerView"
>
  <View
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "flexGrow": undefined,
        "opacity": 1,
      }
    }
    testID="Touchable"
  >
    <View
      style={
        Array [
          Array [
            Object {},
            Object {
              "backgroundColor": "#219EE5",
              "borderRadius": 8,
              "flexDirection": "row",
              "justifyContent": "center",
            },
          ],
        ]
      }
      testID="TouchableView"
    />
  </View>
</View>
`;
